src/vpn-api/utils/clients.ts

```diff
@@ -78,59 +79,56 @@ const setDeviceState = (() => {
 					}
 					// Update the current state on success
 					deviceStates[uuid].currentState = targetState;
-					// Log updated state
-					let stateMsg = `common_name=${targetState.common_name} connected=${
-						targetState.connected
-					}`;
-					if (targetState.virtual_address != null) {
-						stateMsg = `${stateMsg} virtual_address=${
-							targetState.virtual_address
-						}`;
-					}
-					logger.info(`Successfully updated state for ${uuid}: ${stateMsg}`);
+					return targetState;
 				})
 				.catch(err => {
 					captureException(err, 'Error updating state', {
 						user: { uuid },
 					});
 					// Add a 60 second delay in case of failure to avoid a crazy flood
-					return Promise.delay(60000).then(() => {
+					return Bluebird.delay(60000).then(() => {
 						// Trigger another apply, to retry the failed update
 						applyState(uuid);
 						// Since we are recursing and this function always extends
 						// the promise chain (deviceStates[uuid].promise.then ->..)
-						// we need to return undefined to make this promise resolve
+						// we need to return targetState to make this promise resolve
 						// and let it continue with the recursion. If we just
 						// returned applyState() instead, the whole thing would
 						// deadlock
-						return null;
+						return targetState;
 					});
 				});
-		});
-	};
+		}));
 
 	return (state: DeviceState) => {
 		const uuid = state.common_name;
 		if (deviceStates[uuid] == null) {
 			deviceStates[uuid] = {
-				targetState: {},
+				targetState: state,
 				currentState: {},
-				promise: Promise.resolve(),
+				promise: Bluebird.resolve(),
 			};
+		} else {
+			deviceStates[uuid].targetState = state;
 		}
 		deviceStates[uuid].targetState = state;
-		applyState(uuid);
+		return applyState(uuid);
 	};
 })();
 
-export const connected = (data: Partial<DeviceState>) => {
-	data = _.pick(data, 'common_name', 'virtual_address');
-	data.connected = true;
-	return setDeviceState(data as DeviceState);
+export const connected = (data: VpnClientTrustedData) => {
+	const state = {
+		common_name: data.common_name,
+		connected: true,
+		virtual_address: data.ifconfig_pool_remote_ip,
+	};
+	return setDeviceState(state as DeviceState);
 };
 
-export const disconnected = (data: Partial<DeviceState>) => {
-	data = _.pick(data, 'common_name');
-	data.connected = false;
-	return setDeviceState(data as DeviceState);
+export const disconnected = (data: VpnClientDisconnectData) => {
+	const state = {
```

Comment:
I think this is better as
```suggestion
	const state: DeviceState = {
```
and removing the cast below